engine/events/specials.asm

```diff
@@ -435,11 +435,10 @@ RespawnOneOffs:
 	ret
 
 CheckCaughtFlag:
-; Sets the z flag if the player has not caught the mon
+; EngineFlagAction sets the z flag if the player has not caught the mon
 ; input: de = ENGINE_PLAYER_CAUGHT_*
 	ld b, CHECK_FLAG
-	call _EngineFlagAction
-	dec c
+	farcall EngineFlagAction
```

Comment:
You should always turn `call Foo / ret` into `jmp Foo`, and `farcall Foo / ret` into `farjp Foo`. (PC uses `jmp` as a macro to warn when a `jp` can be optimized to `jr`.) Run `utils/optimize.py` to double-check such micro-issues.

Reply:
The more I read this; the more confused I get. I understand that `jmp` should be used instead of `jp` because it can provide an assert warning if it can be optimized to `jr`, but how can you replace a `call Foo / ret`? Aren't `jp`/`jr` used if you don't need to `ret`?

And looking at the `farjp` macro; i'm not sure what the significance of `BANK(\1) | $80` is doing compared to a normal `farcall`. 

Oh, and I ran utils/optimize.py without changing anything.. and it doesn't appear to have found anything wrong with what i'm doing.

I'm not being argumentative here... I love these learning opportunities for myself!

Reply:
If the only thing you do after a call is a ret, then it can be replaced by a jmp. This is known as a "tail call" if you wanna look up more details on it.

But here's my attempted explanation:
Call sets up a "stack frame" (in general computer science. Here the only thing in the frame is the return address)
Ret then takes the latest stack frame and returns to that location.

So for example:

```
start:
  call A // 1. Set up stack frame #1 and jump to A
  nop // Stack Frame #1 points here. 5. We get here!
  // END
  
A:
  call B // 2. Set up stack frame #2 and jump to B
  ret // Stack frame #2 points here. 4. pop stack frame #1 and jump to start's nop
  
B: 
  ret // 3. Pop stack frame #2 and jump to A's ret
```

I tried to label what happens at each instruction here, in numbered order, hope you can follow!
The tail call optimisation says that stack frame #2 is useless! Why not just jump straight from B back to start, as we don't need to do anything more at A.

Since all call does is setup the stack and jump, we can just jump to skip setting up the stack.
And because we jump out of A, we don't need to ret any more either as we would never reach that instruction.

Sorry if I'm stepping on your toes in your own repo Rangi!

Reply:
Essentially consider `call` == `push pc; jp`, and `farcall `== `push pc, farjp`, while `ret` == `pop pc`.

This means that an unoptimised tailcall is equivalent to:
```
push pc
jp
push pc
jp
//do work
pop pc 
pop pc
```

The inner push/pop isn't needed as you never use the value you're saving.
The same applies if you push/poping registers that don't actually need to be saved.